source/brailleDisplayDrivers/albatross/driver.py

```diff
@@ -0,0 +1,624 @@
+					self._dev = None
+				log.info(
+					f"Connection to {self.description} display on {portType} port {port} "
+					f"at {self._baudRate} bps failed."
+				)
+			if self.numCells:
+				break
+		else:
+			raise RuntimeError("No Albatross found")
+
+	def _initConnection(self) -> bool:
+		"""_initConnection, _initPort, _openPort, _readInitByte and _readSettingsByte
+		are helper functions to establish connection.
+		If no connection, Albatross sends continuously INIT_START_BYTE
+		followed by byte containing various settings like number of cells.
+		"""
+		for i in range(MAX_INIT_RETRIES):
+			if not self._dev:
+				if not self._initPort(i):
+					continue
+			elif not self._dev.is_open:
+				if not self._openPort(i):
+					continue
+			if self._tryToConnect:
+				self._tryToConnect = False
+			else:
+				log.debug(
+					f"Sleeping {SLEEP_TIMEOUT} seconds before try {i + 1} / {MAX_INIT_RETRIES}"
+				)
+				time.sleep(SLEEP_TIMEOUT)
+			if not self._readInitByte():
+				continue
+			if not self._tryToConnect and self.numCells:
+				return True
+		return False
+
+	def _initPort(self, i: int = MAX_INIT_RETRIES - 1) -> bool:
+		"""Initializes port.
+		@param i: Just for loggint retries.
+		"""
+		try:
+			self._dev = serial.Serial(
+				self._currentPort, baudrate=self._baudRate, stopbits=serial.STOPBITS_ONE,
+				parity=serial.PARITY_NONE, timeout=READ_TIMEOUT, writeTimeout=WRITE_TIMEOUT
+			)
+			log.debug(f"Port {self._currentPort} initialized")
+			if not self._resetBuffers():
+				if i == MAX_INIT_RETRIES - 1:
+					return False
+				log(
+					f"sleeping {SLEEP_TIMEOUT} seconds before try {i + 2} / {MAX_INIT_RETRIES}")
+				time.sleep(SLEEP_TIMEOUT)
+				return False
+			return True
+		except IOError:
+			if i == MAX_INIT_RETRIES - 1:
+				log.debug(f"Port {self._currentPort} not initialized", exc_info=True)
+				return False
+			log.debug(
+				f"Port {self._currentPort} not initialized, sleeping {SLEEP_TIMEOUT} seconds "
+				f"before try {i + 2} / {MAX_INIT_RETRIES}", exc_info=True
+			)
+			time.sleep(SLEEP_TIMEOUT)
+			return False
+
+	def _openPort(self, i: int = MAX_INIT_RETRIES - 1) -> bool:
+		"""Opens port.
+		@param i: Just for loggint retries.
+		"""
+		try:
+			self._dev.open()
+			log.debug(f"Port {self._currentPort} opened")
+			if not self._resetBuffers():
+				if i == MAX_INIT_RETRIES - 1:
+					return False
+				log(
+					f"sleeping {SLEEP_TIMEOUT} seconds before try {i + 2} / {MAX_INIT_RETRIES}")
+				time.sleep(SLEEP_TIMEOUT)
+				return False
+			return True
+		except IOError:
+			if i == MAX_INIT_RETRIES - 1:
+				log.debug(f"Port {self._currentPort} not opened", exc_info=True)
+				return False
+			log.debug(
+				f"Port {self._currentPort} not opened, sleeping {SLEEP_TIMEOUT} seconds "
+				f"before try {i + 2} / {MAX_INIT_RETRIES}", exc_info=True
+			)
+			time.sleep(SLEEP_TIMEOUT)
+			return False
+
+	def _readInitByte(self) -> bool:
+		# Strange but very rarely in_waiting causes exception.

```

Comment:
what exception? why?

Reply:
> what exception? why?

From serialwin32.py: there is in in_waiting code:

        if not win32.ClearCommError(self._port_handle, ctypes.byref(flags), ctypes.byref(comstat)):
            raise SerialException("ClearCommError failed ({!r})".format(ctypes.WinError()))